### ISSUE TITLE
Pair list decreasing over time

### ISSUE THREAD

**Issue by epigramx** [freqtrade/freqtrade#4797]
This was mentioned here and there before but it deserves a listing since I keep seeing it, and I keep seeing it reported by others e.g.  from someone here yesterday https://discord.com/channels/700048804539400213/702584639063064586/835282320520773673

![image](https://user-images.githubusercontent.com/2960257/115974907-d7c31400-a568-11eb-897d-a8562f480e5b.png)

but I see no satisfactory answer yet on that.

## Describe your environment

  * Operating system: linux dedicated server
  * Python Version: 3.8.5
  * CCXT version: 1.48.22
  * Freqtrade Version: develop of the last few days and for weeks now

## The problem

Basically in some cases the pair listing becomes smaller in the middle of trading for no apparent reason why (most probably at the point of renewal), and then it may become full again, after the refresh of certain filters renew it again.

In my case I suspect what is going on is related to a volume filter followed by other filters and then a volume filter again that when both volume filterrs are re-applied in the end they sometimes reduce it.

In practice in the log it appears to first report "searching for # pairs" - once - then doing the other filters and then doing "searching for # pairs" - **three times** - which is probably the point of reduction.

**Comment by epigramx** [comment]
Looking at it more closely in the log it appears that the symptom and process that causes it is quite clear:

Step 1: The whole thing does its regular job of volume filtering + other filters + volume filtering again and it gets X number of pairs (correctly)

Step 2: A few minutes later **some of the filters and not all them** had their time to refresh and they do that in isolation of the rest

Step 3: Since the new filters may be now with a different listing they may reduce the previous result.

Basically the problem appears to be that in Step 2 + 3 it does not do the "the full cycle" of "big volume list + other filters + reduced volume list" but it only applies to the reduced list of the final step of the previous full cycle so it reduces the full list until the next full renewal.

**Comment by epigramx** [comment]
Note: the practical cause appears to be that, certain filters like age or volume might be renewing their own listing (because of their refresh interval) without triggering a full cycle of filtering with all filters like it is regularly happening at the launch or in most renewal intervals.

Maybe the filtering code is missing a trigger to cause a full filtering using the full filter list when a single filter refreshes itself (or a particular filter might be missing that part in the code and the rest are fine).

**Comment by xmatthias** [comment]
first of all - please use the edit function if you need to add something to your previous post. 
Doing 3 answers within 1 hour will only clutter the issue and make it very hard to try and help.

then:

* Please provide the full pairlist config (and stake_amount + exchange) with which you are experiencing this.
* Please provide the full log of the affected timeperiod (including which pairs disappeared from beginning to end if possible).

All pairs that are removed from the pairlist will be logged - so the logs will contain which pairlist removed the pairs. Depending on the filters (and configuration), it can be normal that a pairlist becomes smaller / enlarges again over time.

**Comment by epigramx** [comment]
I already did part of that because I have scripts that analyze differences of pair lists which led to more specifics in the replies (I didn't edit it because I considered the replies different/new information). I could describe it in further detail and by doing it I notice further discrepancies (notice the bold words):

At XX:47:39: the moment comes for filtering (some regular spam about "Market is not active" is happening for some pairs)
At XX:47:39: the 1st volume filter reports all its unfiltered pairs (I call "1st vol filter" the one with the most pairs on its "Searching # pairs")
At XX:47:39: the spread filter eliminates some pairs
At XX:47:39: the price filter eliminates some pairs
At XX:47:39: the second volume filter reports the reduced list. **The total reduced number is correct and the order is same but one pair in the middle of the list is missing for no apparent reason**.
[the config only had these filters]

At XX:47:44: (**5 seconds later**) **the 1st volume filter kicks in again**(the one with the highest number on the "Searching # pairs") and it reports the reduced list again of the previous step without any change (correct number of pairs) (I never understood why it did that since that filter is supposed to be the first in the list).

At XX:48:09: (**25 seconds later**) the 1st volume filter kicks in again and reports again the same reduced list. The number of pairs is the same and the pairs are the same **but their order has changed** (I guess it got a new list from the server and by random chance it didn't lose a pair but it changed their order).

At XX:48:09 (same time): **the spread filter kicks in and removes 2 pairs** (I bet we found the culprit or we're close)

At XX:48:09 (same time): the 2nd volume filter (small "Searching # pairs"): reports a list **that is now 2 pairs lower than its goal** (bingo!).

And finally at XX:48:14 (5 seconds later): the 1st/big volume filter kicks in: reports the exact same list of the previous step (and I've no reason why that filter is even reported)

## Conclusion

The spread filter kicked in after the regular full filtering and reduced the list. The volume filtering did not expand the list to add extras to replace the ones removed which is what one would expect. I understand the logging might not express exactly the sequence of events in code (e.g. the spread filter might not be the actual trigger) but that's how it looks.

**Comment by xmatthias** [comment]
if you find the conclusion, please provide a fix as part of a Pull request.

If you expect someone else to do further analysis or provide a fix, then provide the information requested above.

I think i've been very explicit in what i would want in my previous comment, but let me be more explicit:
If you can't or won't provide the information requested above (Pairlist configuration used, logs), i'll have to close this issue as "can't reproduce" - simply because you seem unable  / unwilling to provide the information i explicitly requested.

**Comment by epigramx** [comment]
A pair list config that can cause the issue.

```
			{
				"method": "VolumePairList",
				"number_assets": 45,
				"sort_key": "quoteVolume",
				"refresh_period": 900
			},

			{"method": "AgeFilter", "min_days_listed": 10},

			{"method": "SpreadFilter", "max_spread_ratio": 0.005},

			{"method": "PriceFilter", "low_price_ratio": 0.001},

			{
				"method": "VolumePairList",
				"number_assets": 25,
				"sort_key": "quoteVolume",
				"refresh_period": 900
			}
```
By the way notice that the issue is relatively rare (it appears to kick in when a spread filter for some reason kicks in after the regular filtering) but it does happen occasionally in a single day once or twice at least.

The age filter didn't happen to filter out anything in that case.

**Comment by xmatthias** [comment]
to be fair, i don't think `AgeFilter` does  filter pairs often anyway (and it always filters the same pair for the first X days).

With the provided config section, i was able to copy/paste a configuration, recreate the problem and fix it pretty quickly.

**Comment by epigramx** [comment]
> to be fair, i don't think `AgeFilter` does filter pairs often anyway (and it always filters the same pair for the first X days).
> 
> With the provided config section, i was able to copy/paste a configuration, recreate the problem and fix it pretty quickly.

So what was happening in practice? Every single pairlist should have a cache and it was erroneously caching globally, and it was wrong because volume filter specifically also acts as a sort-of "initial whitelist"?

I'm singling out volume filter because I noticed it's the only file that is edited that it's not a global filter file.

**Comment by xmatthias** [comment]
the only pairlists that need a cache (or some way to regenerate the original pairlist) is the pairlist generators (Static and Volume).

Static was never a problem, as it simply gives you pair_whitelist.
Volume however did not cache it's own output, but the final result of all pairlist filters - so if you had 45 in the first volume, and reduce this to 25 in the last volumefilter, the first volumefilter only ever produced 25 pairs until the next cache refresh.

it's not a problem as long as the filters are stable (don't change pairs around semi-randomly) ... but obviously with filters like spreadfilter (hwich can change a lot), it becomes a different topic.

**Comment by epigramx** [comment]
> the only pairlists that need a cache (or some way to regenerate the original pairlist) is the pairlist generators (Static and Volume).
> 
> Static was never a problem, as it simply gives you pair_whitelist.
> Volume however did not cache it's own output, but the final result of all pairlist filters - so if you had 45 in the first volume, and reduce this to 25 in the last volumefilter, the first volumefilter only ever produced 25 pairs until the next cache refresh.
> 
> it's not a problem as long as the filters are stable (don't change pairs around semi-randomly) ... but obviously with filters like spreadfilter (hwich can change a lot), it becomes a different topic.

Thanks. Yes: it was happening quite a lot because it seems spread filter is only based on ticker data that are very up to date. 

What I still don't get is why since we set up "filter 1->2->3->4->5" in the config: it didn't always enforce the full sequence of the filters when one of them changes (so that the initial problem wouldn't occur) but it could do later only volume + spread on their own and leave it at that.
